crates/ruff_linter/src/rules/ruff/rules/unused_variable_accessed.rs

```diff
@@ -0,0 +1,124 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_python_semantic::Binding;
+use ruff_python_stdlib::builtins::is_python_builtin;
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for usages of variables marked as unused (variable names starting with an underscore, except '_') in functions.
+/// Forbid using method and function variables that are marked as unused.
+///
+/// ## Why is this bad?
+/// Marking variables with a leading underscore conveys that they are intentionally unused within the function or method.
+/// When these variables are later referenced in the code, it causes confusion and potential misunderstandings about
+/// the code's intention. A variable marked as "unused" but subsequently used suggests oversight or unintentional use
+/// and detracts from the clarity and maintainability of the codebase.
+///
+/// ## Example
+/// ```python
+/// def function():
+///     _variable = 3
+///     return _variable + 1
+/// ```
+///
+/// Use instead:
+/// ```python
+/// def function():
+///     variable = 3
+///     return variable + 1
+/// ```
+#[derive(ViolationMetadata)]
+pub(crate) struct UnusedVariableAccessed {
+    name: String,
+    // fix: String,
+    shadowed_kind: ShadowedKind,
+}
+
+impl Violation for UnusedVariableAccessed {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!(
+            "Local variable `{}` marked as unused is accessed",
+            self.name
+        )
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some(match self.shadowed_kind {
+            ShadowedKind::BuiltIn => {
+                "Consider using preferred trailing underscores to avoid shadowing a built-in."
+                    .to_string()
+            }
+            ShadowedKind::Some => {
+                "Consider using preferred trailing underscores to avoid shadowing a variable."
+                    .to_string()
+            }
+            ShadowedKind::None => "Consider removing leading underscores.".to_string(),
+        })
+    }
+}
+
+/// RUF052
+pub(crate) fn unused_variable_accessed(
+    checker: &Checker,
+    binding: &Binding,
+) -> Option<Vec<Diagnostic>> {
+    let name = binding.name(checker.source());
+
+    // only used variables
+    if !name.starts_with('_')
+        || name == "_"
```

Comment:
we should also exclude dunder variables from this check, I think. You can use this function to determine if a variable has a dunder name or not: https://github.com/astral-sh/ruff/blob/6f1cf5b686e67fb74383641fcbbe98e33481a271/crates/ruff_python_ast/src/helpers.rs#L547-L549